managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java

```diff
@@ -1356,6 +1356,109 @@ public void terminateFailed(ManagedLedgerException exception, Object ctx) {
         return result;
     }
 
+    @Override
+    public CompletableFuture<Void> asyncAddLedgerProperty(long ledgerId, String key, String value) {
+        CompletableFuture<Void> f = new CompletableFuture<>();
+        // Execute the operation in ML executor to avoid concurrent modifications to ledger info.
+        if (state.isFenced()) {
+            f.completeExceptionally(new ManagedLedgerFencedException());
+            return f;
+        }
+        LedgerInfo li = ledgers.get(ledgerId);
+        if (li == null) {
+            f.completeExceptionally(new ManagedLedgerException("Ledger not found"));
+            return f;
+        }
+
+        synchronized (this) {
+            if (!metadataMutex.tryLock()) {
+                // retry in 100 milliseconds
+                scheduledExecutor.schedule(
+                        () -> asyncAddLedgerProperty(ledgerId, key, value), 100,
+                        TimeUnit.MILLISECONDS);
+            } else {
+                f.whenComplete((v, t) -> metadataMutex.unlock());
+                List<MLDataFormats.KeyValue> oldProperties = li.getPropertiesList();
+                Map<String, String> newPropertiesMap = new HashMap<>();
+                oldProperties.forEach(kv -> newPropertiesMap.put(kv.getKey(), kv.getValue()));
+                newPropertiesMap.put(key, value);
+                updateAndPersistLedgerProperties(newPropertiesMap, li, f);
+            }
+        }
+        return f;
+    }
+
+    @Override
+    public CompletableFuture<Void> asyncRemoveLedgerProperty(long ledgerId, String key) {
+        CompletableFuture<Void> f = new CompletableFuture<>();
+        if (state.isFenced()) {
+            f.completeExceptionally(new ManagedLedgerFencedException());
+            return f;
+        }
+        LedgerInfo li = ledgers.get(ledgerId);
+        if (li == null) {
+            f.completeExceptionally(new ManagedLedgerException("Ledger not found"));
+            return f;
+        }
+
+        synchronized (this) {
+            if (!metadataMutex.tryLock()) {
+                // retry in 100 milliseconds
+                scheduledExecutor.schedule(
+                        () -> asyncRemoveLedgerProperty(ledgerId, key), 100,
+                        TimeUnit.MILLISECONDS);
+            } else {
+                f.whenComplete((v, t) -> metadataMutex.unlock());
+                List<MLDataFormats.KeyValue> oldProperties = li.getPropertiesList();
+                Map<String, String> newPropertiesMap = new HashMap<>();
```

Comment:
We can optimize this by adding a fast path: if the properties don't contain the given key, we can skip rebuilding and return early.